### Que1. What is the difference between a function and a method in Python?

The key difference between a function and a method in Python lies in their association with objects and how they are called. Here's a breakdown:

#### 1. Function:
##### Definition: 
A function is a block of reusable code that is not associated with any particular object. It can be called independently to perform a task.
##### Declaration: 
Defined using the def keyword.
##### Usage: 
Functions can take arguments (parameters) and return values.

##### Example:

In [2]:
def greet(name):
         return f"Hello, {name}!"

In [7]:
greet("Aalind")

'Hello, Aalind!'

#### 2. Method:
##### Definition: 
A method is a function that is associated with an object (typically part of a class). It operates on the data (attributes) of the object and is called using the object.
##### Binding:
Methods are functions that are bound to objects, meaning they can modify or interact with the object’s state.
##### Types of Methods:
- Instance Methods: Operate on the instance of the class and can access or modify object attributes.
- Class Methods: Bound to the class rather than instances and can modify class-level attributes.
- Static Methods: Do not modify object or class state but belong to a class for organizational purposes.

##### Example

In [46]:
class Greeter:
         def __init__(self, name):
             self.name = name
         
         def greet(self):
             return f"Hello, {self.name}!"
     
g = Greeter("Jack")
print(g.greet())

Hello, Jack!


### Key Differences:
| *Aspect*              | *Function*                               | *Method*                               |
|-------------------------|--------------------------------------------|------------------------------------------|
| *Association*          | Independent, not associated with objects   | Associated with an object (usually a class instance) |
| *Calling*              | Called directly, e.g., function()        | Called on an object, e.g., object.method() |
| *Data Access*          | Cannot directly access or modify object attributes | Can access or modify the object's attributes (instance methods) |
| *Declaration Context*  | Defined at the module or global level      | Defined inside a class                   |

In summary, functions are independent units of code, while methods are functions bound to objects and are typically used to operate on an object’s attributes or state.

### Que2. Explain the concept of function arguments and parameters in Python.

In Python, function arguments and parameters refer to the values and variables involved when calling and defining functions. Though often used interchangeably, they represent different aspects of how information is passed into a function.

#### 1. *Parameters*:
##### Definition: 
Parameters are the variables defined by the function that receive input values. These are specified in the function definition.
##### Role: 
Parameters act as placeholders for the values that will be passed to the function when it is called.

##### Example:

In [47]:
def greet(name):  # 'name' is a parameter
         return f"Hello, {"Jack"}!"

In [48]:
greet("name")

'Hello, Jack!'

#### 2. Arguments:
##### Definition: 
Arguments are the actual values passed to the function when it is called. These values are assigned to the corresponding parameters defined in the function.
##### Role:
Arguments provide the actual data that the function will use during its execution.

##### Example:

In [49]:
greet("Jack")  # "Alice" is an argument passed to the function

'Hello, Jack!'

#### Key Types of Function Parameters and Arguments:

##### A. Positional Arguments/Parameters:
Arguments are matched to parameters based on their position in the function call.

##### Example:

In [51]:
def add(a, b):  # 'a' and 'b' are parameters
         return a + b

In [53]:
add(5, 7)  # 5 and 7 are arguments, matched to 'a' and 'b'

12

##### B. Keyword Arguments:
Arguments are passed to the function by explicitly specifying the parameter name. This allows you to pass arguments in any order.

##### Example:

In [73]:
def introduce(name, age):
         return f"{name} is {age} years old."
     
introduce(name="Jack", age= 26)

'Jack is 26 years old.'

##### C. Default Parameters:

Parameters can be given default values, which are used if no argument is provided for that parameter during the function call.

##### Example:

In [77]:
def greet(name="Guest"):
         return f"Hello, {name}!"
     
greet()          # Output: "Hello, Guest!"
greet("Jack")   # Output: "Hello, Jack!"

'Hello, Jack!'

##### D. Variable-Length Arguments:
Python allows you to pass a variable number of arguments to a function using args and kwargs.

##### 1. args (Positional variable-length arguments):
Allows the function to accept any number of positional arguments, which are then stored as a tuple.
##### Example:

In [78]:
 def sum_numbers(*args):
            return sum(args)
        
sum_numbers(1, 2, 3)

6

##### 2. kwargs (Keyword variable-length arguments):
Allows the function to accept any number of keyword arguments, which are stored in a dictionary.
#####Example:

In [81]:
def print_info(**kwargs):
            for key, value in kwargs.items():
                print(f"{key}: {value}")
        
print_info(name="Jack", age=26, city="New York")

name: Jack
age: 26
city: New York


#### Summary:
##### Parameters: 
Defined in the function definition, they are placeholders for the values that will be passed to the function.
##### Arguments:
The actual values passed into the function when it is called.

Together, parameters and arguments allow functions to operate with flexible inputs, making them powerful tools in Python programming.

### Que3. What are the different ways to define and call a function in Python?

In Python, functions are defined and called in different ways depending on the context. Below are some common methods:

#### 1. Basic Function Definition and Call:
A function is defined using the def keyword followed by a name, parameters (optional), and the function body. You call the function by its name followed by parentheses.

##### Defining a function

In [2]:
def greet():
    print("Hello, World!")

##### Calling the function

In [3]:
greet() 

Hello, World!


#### 2. Function with Parameters:
A function can take arguments that are passed when calling the function

##### Defining a function with parameters

In [6]:
def greet(name):
    print(f"Hello, {name}!")

##### Calling the function with an argument

In [7]:
greet("Jack")

Hello, Jack!


#### 3. Function with Default Parameters:
You can provide default values for parameters.

##### Defining a function with a default parameter

In [9]:
def greet(name="World"):
    print(f"Hello, {name}!")

##### Calling without arguments uses the default value

In [10]:
greet()

Hello, World!


##### Calling with an argument overrides the default

In [11]:
greet("Jack")

Hello, Jack!


#### 4. Function with Return Values:
A function can return a value using the return statement.

##### Defining a function that returns a value

In [12]:
def add(a, b):
    return a + b

##### Calling the function and storing the result

In [13]:
result = add(2, 3)
print(result)

5


#### 5. Lambda Functions (Anonymous Functions)
A lambda function is a short, inline function without a name. It is defined using the lambda keyword and can take multiple arguments but contains only one expression.

##### Defining a lambda function

In [17]:
square = lambda x: x ** 3

##### Calling the lambda function

In [18]:
print(square(4))

64


##### Lambda with multiple parameters

In [19]:
add = lambda a, b: a + b
print(add(3, 4))

7


#### 6. Higher-Order Functions:
A function that takes another function as an argument or returns one.


##### Defining a higher-order function


In [22]:
def apply_func(f, x):
    return f(x)

##### Passing a lambda function to the higher-order function

In [23]:
result = apply_func(lambda x: x ** 2, 5)
print(result)

25


#### 7. Recursive Functions:
A function that calls itself.

##### Defining a recursive function to calculate factorial

In [25]:
def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n - 1)

##### Calling the recursive function

In [27]:
print(factorial(5))

120


#### 8. Arbitrary Arguments:
You can define a function to accept any number of positional (*args) or keyword arguments (**kwargs).

##### Defining a function with *args (arbitrary positional arguments)

In [29]:
def add_all(*args):
    return sum(args)

##### Calling with multiple arguments

In [31]:
print(add_all(1, 2, 3, 4))

10


##### Defining a function with **kwargs (arbitrary keyword arguments)

In [33]:
def greet_all(**kwargs):
    for key, value in kwargs.items():
        print(f"{key}: {value}")

In [34]:
# Calling with multiple keyword arguments
greet_all(name="Jack", age=26)

name: Jack
age: 26


These are some of the common ways to define and call functions in Python, each suited to different use cases and contexts.

### Que4. what is the purpose of the 'return' statment in a Python function?

The 'return' statement in a Python function is used to:

#### 1. Send a Result Back to the Caller: 
It terminates the execution of the function and sends a value (or values) back to the part of the program where the function was called. This allows the calling code to capture and use the result of the function.

  ##### Example:

In [38]:
def add(a, b):
       return a + b

result = add(4, 7)  # 'result' will hold the value 5
print(result)

11


#### 2. Exit the Function:
Once the return statement is executed, the function terminates, meaning no further lines of code within the function will be executed.

   ##### Example:

In [41]:
 def check_even(number):
       if number % 2 == 0:
           return "Even"
       return "Odd"

print(check_even(4))  
print(check_even(3))

Even
Odd


#### 3. Return Multiple Values:
Python allows a function to return multiple values as a tuple.

  ##### Example:

In [44]:
def get_coordinates():
       x = 10
       y = 20
       return x, y

In [46]:
coord = get_coordinates()  
print(coord)

(10, 20)


In [47]:
# You can also unpack the returned tuple
x_val, y_val = get_coordinates()
print(x_val, y_val)

10 20


#### 4. Return None by Default: 
If a function does not explicitly use a return statement, or it doesn't return any value, Python implicitly returns None.

  ##### Example:

In [50]:
def greet():
    print("Hello, World!")

In [51]:
result = greet() 

Hello, World!


In [53]:
print(result)

None


#### Key Points:
##### Without return:
The function always returns None.
##### With return:
The function can return any type of object, including integers, strings, lists, objects, or multiple values.
##### Return Ends Function Execution:
Any code after the return statement will not be executed.

### Que 5. what are iterators in python and how do they differ from iterables?

In Python, iterators and iterables are closely related concepts, but they serve different roles in iteration (i.e., looping over items).

#### 1. Iterables:
An *iterable* is any Python object that can return its elements one at a time. It is an object that can be iterated (looped) over. Most commonly used iterables include lists, tuples, strings, dictionaries, and sets.

#### Key Characteristics of Iterables:
- An iterable object has the __iter__() method, which returns an iterator.
- Examples of iterable objects:
- Lists: [1, 2, 3]
- Tuples: (1, 2, 3)
- Strings: "abc"
- Sets: {1, 2, 3}
- Dictionaries: {"key1": "value1", "key2": "value2"}

##### Example of an Iterable:

In [54]:
my_list = [1, 2, 3]  
for item in my_list:
    print(item)

1
2
3


#### 2. Iterators:
An iterator is an object that represents a stream of data. It produces the next value each time it is asked for one, usually with the next() function.

#### Key Characteristics of Iterators:
- An iterator is created by calling the __iter__() method on an iterable.
- An iterator must implement two methods:
1. __iter__() – returns the iterator object itself.
2. __next__() – returns the next element in the sequence or raises a StopIteration exception when there are no more elements.

##### Example of an Iterator:

In [56]:
my_list = [1, 2, 3]
iterator = iter(my_list)  

In [57]:
print(next(iterator))

1


In [58]:
print(next(iterator))

2


In [59]:
print(next(iterator)) 

3


When there are no more elements to return, calling next() again will raise a StopIteration exception:

print(next(iterator))

#### 3. Difference Between Iterables and Iterators:

| *Aspect*         | *Iterable*                                             | *Iterator*                                           |
|--------------------|----------------------------------------------------------|--------------------------------------------------------|
| *Definition*      | An object that can be looped over (e.g., list, tuple).   | An object that produces one element at a time.          |
| *Methods*         | Must implement the __iter__() method.                  | Must implement both __iter__() and __next__() methods. |
| *Usage*           | You can get an iterator from an iterable using iter(). | The iterator is used with next() to get elements one by one. |
| *Reusability*     | An iterable can be passed through multiple loops.        | An iterator can only be used once, after which it’s exhausted. |
| *Memory Efficiency* | Holds all items in memory at once (unless it's a generator or similar). | Does not store all elements in memory; fetches elements one at a time. |


#### 4. Custom Iterators:
You can create custom iterators by defining a class that implements both __iter__() and __next__() methods.

##### Example of a Custom Iterator:

In [79]:
class Counter:
    def __init__(self, start, end):
        self.current = start
        self.end = end

    def __iter__(self):
        return self

    def __next__(self):
        if self.current > self.end:
            raise StopIteration
        else:
            self.current += 1
            return self.current - 1

# Create an instance of the custom iterator
counter = Counter(1, 5)

# Iterate using a for loop
for num in counter:
    print(num) 

1
2
3
4
5


#### 5. Generators (A Type of Iterator):
A generator is a simpler way to create iterators using functions and the yield keyword.

##### Example of a Generator:

In [80]:
def countdown(n):
    while n > 0:
        yield n
        n -= 1

# Get a generator object
gen = countdown(5)

# Iterate through the generator
for num in gen:
    print(num)

5
4
3
2
1


#### Summary:
##### Iterable:
Any object capable of returning its elements one at a time (supports __iter__()).
##### Iterator: 
An object that represents a stream of data and supports both __iter__() and __next__(). It is used to get elements from an iterable one at a time.

### Que 6.  Explain the concept of generators in Python and how they are defined.

In Python, a generator is a special type of iterator that allows you to iterate over a sequence of values lazily, meaning the values are generated on-the-fly as you iterate over them, rather than being stored in memory all at once. This makes generators very memory-efficient, especially when dealing with large datasets or streams of data.

How Generators Work
Generators work by using a yield statement instead of a return statement in a function. When a function contains a yield statement, it automatically becomes a generator function. When the generator function is called, it doesn’t actually execute the function body immediately. Instead, it returns a generator object that can be iterated over.

When you iterate over a generator, the function's code is executed up to the point of the first yield statement. The value specified by yield is then returned to the caller, and the function's state is preserved. The next time the generator is iterated, execution resumes from where it left off, continuing until the next yield is encountered.

Example of a Generator
Here’s an example of a simple generator function that yields a sequence of numbers:

In [2]:
def count_up_to(max):
    count = 1
    while count <= max:
        yield count
        count += 1

You can use this generator as follows:

In [3]:
counter = count_up_to(5)
for number in counter:
    print(number)

1
2
3
4
5


#### Key Points-

Memory Efficiency: Generators are memory-efficient because they generate values one at a time and only as needed. This is particularly useful for handling large data streams or sequences.
Lazy Evaluation: Values are produced on-the-fly and not stored in memory.
Infinite Sequences: Generators are well-suited for representing infinite sequences, like an endless stream of numbers, since they don’t require all values to be computed and stored at once.

Generator Expressions
In addition to generator functions, Python also supports generator expressions, which are similar to list comprehensions but produce a generator instead of a list.

##### Example:

In [13]:
squares = (x * x for x in range(10))

This expression creates a generator that yields the squares of numbers from 0 to 9. You can iterate over it in the same way:

In [14]:
for square in squares:
    print(square)

0
1
4
9
16
25
36
49
64
81


#### Conclusion:
Generators are a powerful tool in Python for creating iterators with less memory overhead and for managing large or infinite sequences of data efficiently. They allow for lazy evaluation, making them ideal for scenarios where you don’t need all values at once.

### Que 7. What are the advantages of using generators over regular functions?

Generators offer several advantages over regular functions, particularly when it comes to memory usage, performance, and code readability. Here are the key benefits:

#### 1. Memory Efficiency

Lazy Evaluation: Generators produce items one at a time and only when needed, rather than generating all items at once and storing them in memory. This is especially useful when dealing with large datasets or streams of data.

Reduced Memory Footprint:Since generators yield items one by one, they don’t require the entire sequence to be stored in memory. This can significantly reduce the memory footprint compared to regular functions that return lists or other collections.


#### 2. Performance

Improved Performance for Large Data: For operations on large datasets, generators can lead to faster execution because they avoid the overhead of loading the entire dataset into memory at once.

Pipelining: Generators can be used in a pipeline, where the output of one generator is fed directly as input to another, enabling efficient data processing in a streaming manner.


#### 3. Infinite Sequences

Handling Infinite Data Streams: Generators are ideal for representing infinite sequences, such as continuous data streams or endless mathematical sequences, because they generate values as needed rather than trying to create an infinite list

#### 4. Simpler Code

Cleaner Code for Iteration: Generators allow you to write more readable and maintainable code for iteration. The use of yield often makes the logic of iteration more explicit and easier to follow.

State Preservation: Generators automatically maintain their state between each iteration, making it easier to implement complex iteration logic without manually managing state variables.

#### 5. Time Efficiency

Faster Start Time: Because a generator produces its first value without generating the entire sequence, you can start processing data faster compared to regular functions that return a fully computed list or sequence.

#### 6. Composability

Pipeline Construction: Generators can be composed easily, allowing you to build pipelines where data is processed in stages. This composability can lead to more modular and reusable code.

Example Comparison

##### Regular Function:

In [15]:
def square_numbers(nums):
    result = []
    for i in nums:
        result.append(i * i)
    return result

squares = square_numbers(range(1000000))

##### Generator Function:

In [18]:
def square_numbers(nums):
    for i in nums:
        yield i * i

squares = square_numbers(range(1000000))

Memory Usage: The regular function creates and returns a list, which requires storing all squared values in memory at once. The generator function, on the other hand, computes each square on-the-fly, significantly reducing memory usage.

Execution Speed: The generator function can start yielding results immediately without waiting for the entire list to be computed, making the program more responsive.

#### Conclusion:

Generators offer significant advantages over regular functions when dealing with large datasets, infinite sequences, or when you need more efficient memory and time management. They lead to more efficient, readable, and maintainable code, making them a valuable tool in Python programming.

### Que 8. What is a lambda function in Python and when is it typically used?

A lambda function in Python is a small, anonymous function defined with the lambda keyword. Unlike regular functions defined using def, lambda functions are limited to a single expression and do not have a name (hence "anonymous"). They are typically used for short, simple operations that are needed temporarily and are not complex enough to require a full function definition.

Syntax of a Lambda Function
The syntax for a lambda function is:

In [19]:
lambda arguments: expression

<function __main__.<lambda>(arguments)>

arguments: A comma-separated list of inputs (just like the arguments in a regular function).

expression: A single expression that is evaluated and returned by the lambda function.

Example of a Lambda Function

Here’s a simple example that demonstrates how a lambda function works:

In [21]:
add = lambda x, y: x + y
result = add(3, 5)
print(result)

8


In this example, lambda x, y: x + y creates a lambda function that takes two arguments, x and y, and returns their sum. The lambda function is then assigned to the variable add, which can be called like a normal function.
Typical Use Cases for Lambda Functions
Lambda functions are typically used in situations where a small, throwaway function is needed for a short period of time. Some common scenarios include:

Using with Functions Like map(), filter(), and reduce():

#### 1.map(): 
Applies a function to all items in an input list (or other iterable).

In [22]:
numbers = [1, 2, 3, 4, 5]
squares = map(lambda x: x * x, numbers)
print(list(squares))

[1, 4, 9, 16, 25]


#### 2.filter(): 
Filters items in an iterable based on a function that returns True or False.

#### 3.reduce(): 
(from functools) Reduces an iterable to a single value using a binary function.

2. Sorting and Ordering with sorted() and sort(): Lambda functions can be used to define custom sorting criteria.

In [23]:
points = [(2, 3), (1, 2), (4, 1), (3, 3)]
sorted_points = sorted(points, key=lambda point: point[1])
print(sorted_points)

[(4, 1), (1, 2), (2, 3), (3, 3)]


#### 3. Inline Callbacks: 
When a function needs to be passed as an argument for callbacks or event handling, lambda functions can be useful.

In [24]:
def apply_function(f, value):
    return f(value)

result = apply_function(lambda x: x ** 2, 5)
print(result)

25


#### 4. Simple Operations in List Comprehensions or Data Structures:
Sometimes, lambda functions are used for quick, simple operations within a list comprehension or to initialize data structures.

In [25]:
multiply_by_two = lambda x: x * 2
numbers = [multiply_by_two(x) for x in range(5)]
print(numbers)

[0, 2, 4, 6, 8]


#### Limitations of Lambda Functions
#####Single Expression: 
Lambda functions are limited to a single expression. They cannot contain statements, so complex logic that requires multiple lines or statements is better handled with a regular function.
No Annotations or Docstrings: Lambda functions are anonymous and do not support function annotations or docstrings, making them less informative and harder to debug.
##### Reduced Readability: 
Overusing lambda functions can make code harder to read, especially for more complex operations.
####Conclusion
Lambda functions are a convenient way to create small, throwaway functions on the fly for simple tasks. They are commonly used in situations where a short function is needed for a brief period, such as with map(), filter(), sorted(), or as inline callbacks. However, for more complex logic or when readability is a concern, it’s usually better to define a regular function.

### Que 9.  Explain the purpose and usage of the `map()` function in Python
The map() function in Python is used to apply a given function to all the items in an input iterable (such as a list, tuple, or set) and return a map object (which is an iterator) containing the results. The primary purpose of map() is to transform data by applying a function to each item in an iterable, making it a convenient tool for processing and transforming collections of data.
Syntax of map()-
map(function, iterable, ...)
#### 1.function: 
The function that you want to apply to each item in the iterable(s). This function should take one or more arguments, depending on how many iterables you pass.
#### 2.Iterable: 
The iterable (e.g., list, tuple) that you want to process.
You can pass multiple iterables if the function requires more than one argument. In this case, map() will apply the function to corresponding items from each iterable in parallel.

##### Example of map() with a Single Iterable
Here’s a simple example where map() is used to square each number in a list:

In [26]:
numbers = [1, 2, 3, 4, 5]
squared_numbers = map(lambda x: x ** 2, numbers)
print(list(squared_numbers))

[1, 4, 9, 16, 25]


##### In this example:

The lambda x: x ** 2 function squares each element.

map() applies this function to each item in the numbers list.

The result is a map object, which is converted to a list to display the results.
Example of map() with Multiple Iterables

##### Example of map() with multiple iterables. 

Here’s an example where two lists are added element-wise:

In [27]:
numbers1 = [1, 2, 3]
numbers2 = [4, 5, 6]
summed_numbers = map(lambda x, y: x + y, numbers1, numbers2)
print(list(summed_numbers))

[5, 7, 9]


##### In this example:

The lambda x, y: x + y function adds corresponding elements from numbers1 and numbers2.

map() processes both lists in parallel, applying the function to each pair of elements.

#### Key Points About map()

##### Returns an Iterator: 
map() returns a map object, which is an iterator. This means that the results are computed lazily and can be iterated over. You often convert the map object to a list, tuple, or other data structure to see the results immediately.

##### Works with Any Iterable: 
map() can work with any iterable, not just lists. This includes tuples, sets, strings, and even custom iterables.

##### Function Must Accept Multiple Arguments If Multiple Iterables Are Provided: 
If you pass multiple iterables to map(), the function must accept that many arguments, and the iterables must be of the same length.

#### Practical Usage of map()

##### Data Transformation: 
map() is commonly used for transforming data. For example, converting temperatures from Celsius to Fahrenheit, or extracting a particular field from a list of dictionaries.

##### Function Application: 
When you need to apply a function to each item in a collection, map() provides a concise and readable way to do this without writing an explicit loop.

##### Element-wise Operations:
map() is useful for performing element-wise operations on multiple sequences, such as adding corresponding elements from two lists.

#### Conclusion

The map() function is a powerful and flexible tool in Python for applying a function to every item in one or more iterables. It allows for clean and efficient data transformation, making it an essential function for data processing tasks. Since map() returns an iterator, it is also memory efficient, especially when dealing with large datasets.


### Que 10 What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?
In Python, map(), reduce(), and filter() are higher-order functions that allow you to process and transform data in iterables like lists, tuples, or sets. Each of these functions serves a different purpose:
#### 1. map() Function

Purpose: 
The map() function applies a given function to each item in an iterable and returns an iterator (a map object) with the results.

Usage: 
It is used when you need to transform or process each element of an iterable independently.

##### Syntax:
map(function, iterable, ...)

##### Example:

In [28]:
numbers = [1, 2, 3, 4, 5]
squares = map(lambda x: x ** 2, numbers)
print(list(squares))

[1, 4, 9, 16, 25]


Here, map() applies the lambda function to square each number in the list.
#### 2. filter() Function

Purpose: 
The filter() function applies a function to each item in an iterable and returns an iterator (a filter object) containing only the items for which the function returns True.

Usage: 
It is used when you want to filter out elements from an iterable based on some condition.

##### Syntax:
filter(function, iterable)

##### Example:

In [29]:
numbers = [1, 2, 3, 4, 5]
even_numbers = filter(lambda x: x % 2 == 0, numbers)
print(list(even_numbers))

[2, 4]


In this case, filter() selects only the even numbers from the list.
#### 3. reduce() Function

Purpose: 
The reduce() function, from the functools module, applies a given function cumulatively to the items of an iterable, from left to right, so as to reduce the iterable to a single value.

Usage: 
It is used when you want to reduce a sequence of elements to a single cumulative value.

##### Syntax:
from functools import reduce
reduce(function, iterable, [initializer])

##### Example:

In [30]:
from functools import reduce
numbers = [1, 2, 3, 4, 5]
sum_of_numbers = reduce(lambda x, y: x + y, numbers)
print(sum_of_numbers)

15


### Que 11.  the internal mechanism for sum operation using  reduce function on this given list:[47,11,42,13]; 

# Practical Questions:

### Que 1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.

A Python function that takes a list of numbers as input and returns the sum of all even numbers:

In [33]:
def sum_of_even_numbers(numbers):
    return sum(num for num in numbers if num % 2 == 0)

#### Example usage:

In [34]:
numbers_list = [1, 2, 3, 4, 5, 6]
print(sum_of_even_numbers(numbers_list))

12


This function uses a list comprehension to filter out the even numbers and then sums them up.

### Que 2. Create a Python function that accepts a string and returns the reverse of that string.

A Python function that accepts a string and returns its reverse:

In [36]:
def reverse_string(s):
    return s[::-1]

#### Example usage:

In [37]:
string = "hello"
print(reverse_string(string))

olleh


In [38]:
name = "Laxman"
print(reverse_string(name))

namxaL


This function uses Python's slicing technique to reverse the string.

### Que 3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.

A Python function that takes a list of integers and returns a new list with the squares of each number:

In [40]:
def square_numbers(numbers):
    return [num ** 2 for num in numbers]

#### Example usage:

In [41]:
numbers_list = [1, 2, 3, 4, 5]
print(square_numbers(numbers_list))

[1, 4, 9, 16, 25]


This function uses a list comprehension to iterate through the input list and square each element.

### Que 4. Write a Python function that checks if a given number is prime or not from 1 to 200.

A Python function that checks if a given number (between 1 and 200) is prime or not:

In [42]:
def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

#### Example usage:

In [43]:
number = 29
print(is_prime(number)) 

True


#### Explanation:
- The function checks if the number n is less than or equal to 1, returning False because primes are greater than 1.
- For numbers greater than 1, it checks divisibility from 2 up to the square root of n (using n**0.5), returning False if a divisor is found.
- If no divisors are found, the number is prime.

##### To check numbers from 1 to 200, you can loop through this range:

In [44]:
for num in range(1, 201):
    if is_prime(num):
        print(f"{num} is a prime number")

2 is a prime number
3 is a prime number
5 is a prime number
7 is a prime number
11 is a prime number
13 is a prime number
17 is a prime number
19 is a prime number
23 is a prime number
29 is a prime number
31 is a prime number
37 is a prime number
41 is a prime number
43 is a prime number
47 is a prime number
53 is a prime number
59 is a prime number
61 is a prime number
67 is a prime number
71 is a prime number
73 is a prime number
79 is a prime number
83 is a prime number
89 is a prime number
97 is a prime number
101 is a prime number
103 is a prime number
107 is a prime number
109 is a prime number
113 is a prime number
127 is a prime number
131 is a prime number
137 is a prime number
139 is a prime number
149 is a prime number
151 is a prime number
157 is a prime number
163 is a prime number
167 is a prime number
173 is a prime number
179 is a prime number
181 is a prime number
191 is a prime number
193 is a prime number
197 is a prime number
199 is a prime number


### Que 5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.

In [45]:
class FibonacciIterator:
    def __init__(self, num_terms):
        self.num_terms = num_terms  # Number of terms to generate
        self.current = 0            # First Fibonacci number
        self.next = 1               # Second Fibonacci number
        self.index = 0              # Tracks the current term count

In [46]:
def __iter__(self):
        return self

In [47]:
def __next__(self):
        if self.index >= self.num_terms:
            raise StopIteration
        fib_number = self.current
        self.current, self.next = self.next, self.current + self.next

In [51]:
def __next__(self):
        if self.index >= self.num_terms:
            raise StopIteration
        fib_number = self.current
        self.current, self.next = self.next, self.current + self.next  # Update to the next Fibonacci number
        self.index += 1
        return fib_number

#### Explanation:
- __init__: Initializes the iterator with the number of terms to generate and sets the first two Fibonacci numbers.
- __iter__: Returns the iterator object itself.
- __next__: Calculates the next Fibonacci number, updates the internal state, and raises StopIteration when the specified number of terms is reached.

In the example usage, it generates the first 10 Fibonacci numbers using a for loop.

### Que 6. Write a generator function in Python that yields the powers of 2 up to a given exponent.
A list of numbers as input and returns the sum of all even numbers in the list:

In [73]:
def sum_of_even_numbers(numbers):
    # Use a list comprehension to filter even numbers and sum them up
    even_sum = sum(num for num in numbers if num % 2 == 0)
    return even_sum

#### Example usage:

In [74]:
numbers = [47, 11, 42, 13, 8, 20]
result = sum_of_even_numbers(numbers)
print("The sum of even numbers is:", result)

The sum of even numbers is: 70


#### Explanation:

List Comprehension: num for num in numbers if num % 2 == 0 generates a list of even numbers from the input list.

sum() Function: The sum() function then adds up all the numbers in this list of even numbers.

#### Example Output:
If you run the example code with the list [47, 11, 42, 13, 8, 20], the output will be:
This result comes from summing the even numbers 42, 8, and 20.
### Que 7.  Implement a generator function that reads a file line by line and yields each line as a string.
Python generator function that reads a file line by line and yields each line as a string:

In [75]:
def read_file_line_by_line(file_path):
    try:
        with open(file_path, 'r') as file:
            for line in file:
                yield line.strip()  # strip() removes the trailing newline character
    except FileNotFoundError:
        print(f"The file {file_path} was not found.")
    except IOError:
        print(f"An error occurred while reading the file {file_path}.")

# Example usage:
file_path = 'example.txt'
for line in read_file_line_by_line(file_path):
    print(line)


The file example.txt was not found.


#### Explanation:

open(file_path, 'r'): Opens the file at file_path in read mode ('r').

##### with statement:
Ensures that the file is properly closed after the block is executed, even if an error occurs.

##### for line in file:
Iterates over the file object, reading it line by line.

##### yield line.strip(): 
Yields each line as a string, with trailing newline characters removed using strip().

#### Example Usage:

Suppose you have a file named example.txt with the following content:
Hello, World!
This is a test file.
It contains several lines.

Running the generator function with this file:

In [78]:
file_path = 'example.txt'
for line in read_file_line_by_line(file_path):
    print(line)

The file example.txt was not found.


#### Example Output:
Hello, World!
This is a test file.
It contains several lines.
This generator function efficiently reads and processes large files, as it doesn't load the entire file into memory at once but instead processes it line by line.
### Que 8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple
You can use a lambda function to sort a list of tuples based on the second element of each tuple as follows:

#### Sample list of tuples
tuples_list = [(1, 3), (4, 1), (5, 2), (2, 4)]

#### Sorting the list based on the second element of each tuple
sorted_list = sorted(tuples_list, key=lambda x: x[1])

print(sorted_list)


#### Explanation:

lambda x: x[1]: This lambda function takes a tuple x as input and returns its second element (x[1]).

sorted(tuples_list, key=lambda x: x[1]): The sorted() function sorts the tuples_list based on the value returned by the lambda function, which is the second element of each tuple.

#### Example Output:

Given the list [(1, 3), (4, 1), (5, 2), (2, 4)], the sorted list will be:

In [81]:
[(4, 1), (5, 2), (1, 3), (2, 4)]

[(4, 1), (5, 2), (1, 3), (2, 4)]

Here, the tuples are sorted in ascending order based on their second elements: 1, 2, 3, and 4.
### Que 9 -Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.
A Python program that uses the map() function to convert a list of temperatures from Celsius to Fahrenheit:

In [82]:
# Function to convert Celsius to Fahrenheit
def celsius_to_fahrenheit(celsius):
    return (celsius * 9/5) + 32

# List of temperatures in Celsius
celsius_temperatures = [0, 20, 37, 100]

# Use map() to apply the conversion to each temperature
fahrenheit_temperatures = map(celsius_to_fahrenheit, celsius_temperatures)

# Convert the map object to a list to see the results
fahrenheit_list = list(fahrenheit_temperatures)

print(fahrenheit_list)

[32.0, 68.0, 98.6, 212.0]


### Que 10 -  Create a Python program that uses `filter()` to remove all the vowels from a given string.
Python program that uses the filter() function to remove all the vowels from a given string:

In [83]:
# Function to check if a character is not a vowel
def is_not_vowel(char):
    vowels = 'aeiouAEIOU'
    return char not in vowels

# Given string
input_string = "Hello, World!"

# Use filter() to remove vowels from the string
result_string = ''.join(filter(is_not_vowel, input_string))

print(result_string)


Hll, Wrld!


#### Explanation:

##### is_not_vowel(char): 
This function checks if a character is not a vowel by verifying that it is not in the string 'aeiouAEIOU'.

filter(is_not_vowel, input_string): The filter() function applies is_not_vowel to each character in the input_string, filtering out any characters that are vowels.

##### ''.join(filter(...)): 
The filter() function returns an iterator, so ''.join() is used to join the remaining characters back into a string.

#### Example Output:

Given the string "Hello, World!", the program will output:

"Hll, Wrld!"

This output is the original string with all vowels removed.
#### Que 11. ) Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this:
#### Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the product of the price per item and the quantity. The product should be increased by 10,- € if the value of the order is smaller than 100,00 €.

#### Write a Python program using lambda and map.

To accomplish this task using lambda and map, we will create a Python program that processes the data and returns a list of 2-tuples with the order number and the total price for each order, applying the additional €10 charge if the order value is less than €100.

In [84]:
# Data
orders = [
    [34587, "Learning Python, Mark Lutz", 4, 40.95],
    [98762, "Programming Python, Mark Lutz", 5, 56.80],
    [77226, "Head First Python, Paul Barry", 3, 32.95],
    [88112, "Einführung in Python3, Bernd Klein", 3, 24.99]
]

# Lambda function to calculate the total price
# The lambda takes an order and returns a tuple (Order Number, Total Price)
calculate_price = lambda order: (order[0], round(order[2] * order[3], 2) if order[2] * order[3] >= 100 else round(order[2] * order[3] + 10, 2))

# Use map to apply the lambda function to each order
final_prices = list(map(calculate_price, orders))

# Output the result
print(final_prices)


[(34587, 163.8), (98762, 284.0), (77226, 108.85), (88112, 84.97)]


#### Explanation:

Data Structure: The orders list contains sublists with the following structure: [Order Number, Book Title and Author, Quantity, Price per Item].

#### Lambda Function:

The lambda function calculate_price is used to compute the total price for each order.

order[2] * order[3]: This calculates the product of the quantity and the price per item.

if order[2] * order[3] >= 100: If the total price is greater than or equal to €100, it leaves the price unchanged; otherwise, it adds €10.

round(..., 2): The round function is used to round the total price to two decimal places.

#### map() Function:
The map() function applies the calculate_price lambda to each sublist in orders.

#### list() Function: 
Converts the map object into a list of tuples.
#### Example Output:

Running the program with the provided data will yield the following output:

[(34587, 163.8), (98762, 284.0), (77226, 108.85), (88112, 84.97)]

This output contains the order numbers and the final prices, with the additional €10 charge applied to the last order, which originally had a total of 

less than €100.